### PR TITLE
Support dynamic watching for various workload types.

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: mingzhou.swx <mingzhou.swx@alibaba-inc.com>

### Ⅰ. Describe what this PR does
Support dynamic watching for various workload types.

### CODE REVIEW

File: pkg/controller/rollout/rollout_controller.go [hunk 1]
```diff
@@ -79,6 +90,13 @@ func (r *RolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, err
 	}
 
+	watched, err := util.AddDynamicWatcher(runtimeController, workloadHandler, watchedSet, rollout.Spec.ObjectRef.WorkloadRef)
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if watched {
+		watchedSet.Insert(util.GetGVKFrom(rollout.Spec.ObjectRef.WorkloadRef).String())
```
Comment (furykerry) [inline pkg/controller/rollout/rollout_controller.go:97]: shall we wait for informer cache syncd before continue reconcile?

File: pkg/controller/batchrelease/batchrelease_controller.go [hunk 2]
```diff
@@ -146,6 +158,13 @@ func (r *BatchReleaseReconciler) Reconcile(ctx context.Context, req ctrl.Request
 
 	klog.Infof("Begin to reconcile BatchRelease(%v/%v), release-phase: %v", release.Namespace, release.Name, release.Status.Phase)
 
+	watched, err := util.AddDynamicWatcher(runtimeController, workloadHandler, WatchedSet, release.Spec.TargetRef.WorkloadRef)
+	if err != nil {
+		return reconcile.Result{}, err
+	} else if watched {
+		WatchedSet.Insert(util.GetGVKFrom(release.Spec.TargetRef.WorkloadRef).String())
```
Comment (furykerry) [inline pkg/controller/batchrelease/batchrelease_controller.go:165]: shall we protect WatchedSet against synchronization problems?

Comment (veophi) [reply pkg/controller/batchrelease/batchrelease_controller.go]: done

File: pkg/controller/rollout/rollout_controller.go (hunk 1 above)
Comment (veophi) [reply pkg/controller/rollout/rollout_controller.go]: done

File: pkg/controller/rollout/rollout_controller.go [hunk 3]
```diff
@@ -79,6 +96,18 @@ func (r *RolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, err
 	}
 
+	modified, err := util.AddDynamicWatcher(runtimeController, workloadHandler, watchedWorkload, rollout.Spec.ObjectRef.WorkloadRef)
```
Comment (FillZpp) [inline pkg/controller/rollout/rollout_controller.go:99]: Do not pass in `watchedWorkload`, you should check if the GVK is watched here. If not, then add the watcher dynamically.

**Pull request merged**